### PR TITLE
Minor additions to create_autonomy

### PR DESCRIPTION
**Pull request by betaupsx86**
Updated create_autonomy to make use of the new additions to libcreate. Cliff sensors exposed as a topic as well as wheeldrops. User may want to use this info to create a better safety controller. (Cliff topic only on Create V_3 not sure what version of Create has them available). PWM Main Motor control via topic (Brush motors).

### CODE REVIEW

File: ca_driver/src/create_driver.cpp [hunk 1]
```diff
@@ -568,10 +578,33 @@ void CreateDriver::publishBumperInfo()
   bumper_pub_.publish(bumper_msg_);
 }
 
+void CreateDriver::publishCliffInfo()
+{
+  cliff_msg_.header.stamp = ros::Time::now();
+
+  if (model_.getVersion() >= create::V_3)
```
Comment (jacobperron) [inline ca_driver/src/create_driver.cpp:585]: Although I have not tested the older protocols myself, I think *libcreate* supports cliff detection for any version.

The signal strengths on the other hand are only supported for `create::V_2` and `create::V_3`.

File: ca_msgs/msg/Cliff.msg [hunk 2]
```diff
@@ -0,0 +1,15 @@
+Header header
+
+# Cliff light sensors (Create 2 only) in order from left to right
+# Value = true if a cliff is detected
+bool is_cliff_left
+bool is_cliff_front_left
+bool is_cliff_front_right
+bool is_cliff_right
+
+## Raw light sensor signals
```
Comment (jacobperron) [inline ca_msgs/msg/Cliff.msg:10]: Please comment these are for Create 2 (Roomba 600 series) or greater.

File: ca_msgs/msg/Cliff.msg [hunk 3]
```diff
@@ -0,0 +1,15 @@
+Header header
+
+# Cliff light sensors (Create 2 only) in order from left to right
```
Comment (jacobperron) [inline ca_msgs/msg/Cliff.msg:3]: Change comment to reflect that cliff detection is supported for Create 1 (Roomba 500 series) or greater

File: ca_msgs/msg/Wheeldrop.msg [hunk 4]
```diff
@@ -0,0 +1,7 @@
+Header header
+
+# Contact sensors
+bool is_left_dropped
+bool is_right_dropped
+
```
Comment (jacobperron) [inline ca_msgs/msg/Wheeldrop.msg:6]: Remove these two blank lines.

File: ca_driver/src/create_driver.cpp [hunk 5]
```diff
@@ -392,7 +402,7 @@ void CreateDriver::publishOdom()
   odom_msg_.twist.twist.linear.y = vel.y;
   odom_msg_.twist.twist.angular.z = vel.yaw;
 
-  // Update covariances
+  // covariances
```
Comment (jacobperron) [inline ca_driver/src/create_driver.cpp:405]: Unnecessary change, please revert.

File: ca_driver/include/create_driver/create_driver.h [hunk 6]
```diff
@@ -66,6 +72,8 @@ class CreateDriver
   void setASCIICallback(const std_msgs::UInt8MultiArrayConstPtr& msg);
   void dockCallback(const std_msgs::EmptyConstPtr& msg);
   void undockCallback(const std_msgs::EmptyConstPtr& msg);
+  void mainMotorCallback(const std_msgs::Float32ConstPtr& msg);  // PWM control 0 (OFF)-1.0 (100% DUTY CYCLE?)
```
Comment (jacobperron) [inline ca_driver/include/create_driver/create_driver.h:75]: Valid values are in the range [-1.0, 1.0].

File: ca_driver/src/create_driver.cpp [hunk 7]
```diff
@@ -222,6 +227,11 @@ void CreateDriver::undockCallback(const std_msgs::EmptyConstPtr& msg)
   robot_->setMode(create::MODE_FULL);
 }
 
+void CreateDriver::mainMotorCallback(const std_msgs::Float32ConstPtr& msg)
+{
+  robot_->setMainMotor(msg->data);
```
Comment (jacobperron) [inline ca_driver/src/create_driver.cpp:232]: Please guard against invalid values (ie. outside the range [-1.0, 1.0]) and output a `ROS_ERROR` accordingly.

**Review by jacobperron — CHANGES_REQUESTED** [review]
Please update the README regarding the new publishers and subscriber. 

Also, make the neccesary changes after making the API changes from [libcreate - PR #29](https://github.com/AutonomyLab/libcreate/pull/29).

Thanks!

File: ca_driver/src/create_driver.cpp (hunk 7 above)
Comment (betaupsx86) [reply ca_driver/src/create_driver.cpp:232]: Absolutely! What do you think of migrating most subscribers in the node to a service based model. Few of them really benefit from being topics (IMO just the velocity command subscriber). Other things like setting LEDs, requesting a song or turning off/on motors are just sporadically used and align better to the notion of service. What do you think about this? We also have the benefit of giving feedback to the requester via the service response if something went wrong (ex: the underlying driver failed for whatever reason). Maybe I can start with one and see if it really makes create_autonomy more flexible/powerful.

File: ca_driver/src/create_driver.cpp (hunk 1 above)
Comment (betaupsx86) [reply ca_driver/src/create_driver.cpp:585]: Yes I thought that too. I just didn't want to break the driver because I wasn't absolutely sure it supported it

File: ca_driver/src/create_driver.cpp (hunk 7 above)
Comment (jacobperron) [reply ca_driver/src/create_driver.cpp:232]: I like the subscribers over services because of slightly less overhead on the user's end, but perhaps it does make more sense semantically for some functions. I think song/LED requests are good candidates for services, but I'm inclined to leave motor commands as topics (ie. wheel, brush, and vacuum motors) as people may make calls in quick succession to those.

File: ca_driver/src/create_driver.cpp (hunk 1 above)
Comment (jacobperron) [reply ca_driver/src/create_driver.cpp:585]: I have the hardware to test `V_2` and `V_3` functionality and will do testing before merging this PR.

**Comment by jacobperron** [comment]
@betaupsx86 Let me know if you have plans to work more on this PR. Otherwise, I will go ahead and finish adding these features.

File: ca_driver/src/create_driver.cpp (hunk 7 above)
Comment (SKT-Peanut) [reply ca_driver/src/create_driver.cpp:232]: @jacobperron How do you get the param to be in range [-1,1]? (msg->data < 1) && (msg->data>-1)?

**Comment by jacobperron** [comment]
Closing since this has been open for a long time without any activity.
